packages/astro/src/runtime/server/render/astro/render.ts

```diff
@@ -165,3 +168,119 @@ async function bufferHeadContent(result: SSRResult) {
 		}
 	}
 }
+
+export async function renderToAsyncIterable(
+	result: SSRResult,
+	componentFactory: AstroComponentFactory,
+	props: any,
+	children: any,
+	isPage = false,
+	route?: RouteData
+): Promise<AsyncIterable<Uint8Array> | Response> {
+  const templateResult = await callComponentAsTemplateResultOrResponse(
+    result,
+    componentFactory,
+    props,
+    children,
+    route
+  );
+  if (templateResult instanceof Response)
+    return templateResult;
+  let renderedFirstPageChunk = false;
+  if (isPage) {
+    await bufferHeadContent(result);
+  }
+
+	// This implements the iterator protocol:
+	// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#the_async_iterator_and_async_iterable_protocols
+	// The `iterator` is passed to the Response as a stream-like thing.
+	// The `buffer` array acts like a buffer. During render the `destination` pushes
+	// chunks of Uint8Arrays into the buffer. The response calls `next()` and we combine
+	// all of the chunks into one Uint8Array and then empty it.
+	
+	let error: Error | null = null;
+	// The `next` is an object `{ promise, resolve, reject }` that we use to wait
+	// for chunks to be pushed into the buffer.
+  let next = promiseWithResolvers<void>();
+  const buffer: Uint8Array[] = []; // []Uint8Array
+
+  const iterator = {
+    async next() {
+      await next.promise;
+
+			// If an error occurs during rendering, throw the error as we cannot proceed.
+			if(error) {
+				throw error;
+			}
+
+      // Get the total length of all arrays.
+      let length = 0;
+			for(let i = 0, len = buffer.length; i < len; i++) {
+				length += buffer[i].length;
+			}
+
+      // Create a new array with total length and merge all source arrays.
+      let mergedArray = new Uint8Array(length);
+      let offset = 0;
+			for(let i = 0, len = buffer.length; i < len; i++) {
+				const item = buffer[i];
+        mergedArray.set(item, offset);
+        offset += item.length;
+			}
+
+      // Empty the array. We do this so that we can reuse the same array.
+      buffer.length = 0;
+
+      const returnValue = {
+				// The iterator is done if there are no chunks to return.
+        done: length === 0,
+        value: mergedArray
+      };
+
+      return returnValue;
+    }
+  };
+
+  const destination: RenderDestination = {
+    write(chunk) {
```

Comment:
Would you be open to a PR that makes this fast in Bun as well? 

It would use `[Symbol.asyncIterator]` and you can yield either `ArrayBufferView` or string literals, so you can skip the encode/decode step (if I'm reading this code correctly). This would make faster and use a lot less memory.

Then the iterator function would essentially be:

```ts
// earlier
let { promise, resolve, reject } = Promise.withResolvers();
let pendingChunks = [];

// inside the iterator
async function* iterator() {
  while (true) {
    await promise;
    ({ promise, resolve, reject } = Promise.withResolvers());
    const chunks = pendingChunks;
    pendingChunks = [];
    yield* chunks;
    if (chunks.length === 0) return;
  }
}

const destination: RenderDestination = {
   write(chunk) {
       // ....
```

Reply:
Thanks, assuming this would need to conditionally run only in Node we'd like to hold off on that for now. I didn't particularly want to do the change in this PR but the difference was too big to ignore.

Reply:
I've not looked into performance, but seems you could even return an [async generator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/AsyncGenerator) of strings directly. I don't see why you would need to construct the async iterator yourself, and why you need to mess around with that buffer at all, but perhaps I'm missing something? Returning the generator would be a lot less code, and the following works in Node.js:

```js
async function* renderToAsyncIterable() {
  yield await Promise.resolve('a');
  yield await Promise.resolve('b');
  yield await Promise.resolve('c');
}

const res = new Response(renderToAsyncIterable())
await res.text()
```

Note that you can also use [`for await...of`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of) to iterate over async iterables when needed.

Reply:
In my past refactors, I found that generators tend to be slow. So I think we should avoid it even if it may be easier to read.

Reply:
@bluwy right... seems like generators are 1.8-2x slower than doing it manually (at least in node.js), but seems that they scale equally. But that still leaves the question why not to return a `AsyncIterable<string | Uint8Array>` and avoid creating another buffer.